pkg/proctree/datasource.go

```diff
@@ -0,0 +1,210 @@
+package proctree
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/aquasecurity/tracee/types/datasource"
+	"github.com/aquasecurity/tracee/types/detect"
+)
+
+// DataSource is an implementation to detect.Datasource interface, enveloping the ProcessTree type.
+// It exposes all the relevant information of the tree using the interface methods.
+type DataSource struct {
+	procTree *ProcessTree
+}
+
+func NewDataSource(processTree *ProcessTree) *DataSource {
+	return &DataSource{procTree: processTree}
+}
+
+func (ptds *DataSource) Get(key interface{}) (map[string]interface{}, error) {
+	switch typedKey := key.(type) {
+	case datasource.ProcKey:
+		process, found := ptds.procTree.GetProcessByHash(typedKey.Hash)
+		if !found {
+			return nil, detect.ErrDataNotFound
+		}
+		return map[string]interface{}{
+			"process_info": ptds.exportProcessInfo(process, typedKey.Time),
+		}, nil
+	case datasource.ThreadKey:
+		thread, found := ptds.procTree.GetThreadByHash(typedKey.Hash)
+		if !found {
+			return nil, detect.ErrDataNotFound
+		}
+		return map[string]interface{}{
+			"thread_info": ptds.exportThreadInfo(thread, typedKey.Time),
+		}, nil
+	case datasource.LineageKey:
+		process, found := ptds.procTree.GetProcessByHash(typedKey.Hash)
+		if !found {
+			return nil, detect.ErrDataNotFound
+		}
+		return map[string]interface{}{
+			"process_lineage": ptds.exportProcessLineage(process, typedKey.Time, typedKey.MaxDepth),
+		}, nil
+	default:
+		return nil, detect.ErrKeyNotSupported
+	}
+}
+
+func (ptds *DataSource) Keys() []string {
+	return []string{"datasource.ProcKey", "datasource.ThreadKey", "datasource.LineageKey"}
+}
+
+func (ptds *DataSource) Schema() string {
+	schemaMap := map[string]string{
+		"process_info":    "datasource.ProcessInfo",
+		"thread_info":     "datasource.ThreadInfo",
+		"process_lineage": "datasource.ProcessLineage",
+	}
+	schema, _ := json.Marshal(schemaMap)
+	return string(schema)
+}
+
+func (ptds *DataSource) Version() uint {
+	return 1
```

Comment:
the datasource, its type and the signature should walk together, right ? we should make a constant and put on top of those files with a good comment explaining any change should bump the version to a specific format.

I usually like always-increasing versions such as "202309202130" (YYYYMMDDhhmm of the last change). But its not mandatory and you guys will be the one bumping this, so feel free.

Reply:
Hmm good question.
Your suggested versioning is much clearer.
The only issue is that the containers datasource is using version "1" right now, so I tried to fit the convention.
If we agree that this is preferable than I will also change the version of the containers datasource.

Reply:
I close it as the current versioning does not support semantic versioning.
It will require future refactor.